New subgroups - full changes set

gwawiork opened the pull request:
These changes are related the PR #834. This PR include all changes that were developed by my to implement new subgroups extensions for OpenCL 3.0. 
The first [commit ](https://github.com/KhronosGroup/OpenCL-CTS/commit/a7291f612fc96b72b326081b0ed9572c04583a8a) in this PR covers previous PR #834

polsenferreira commented on test_conformance/subgroups/test_workgroup_extended_types.cpp:
```diff
@@ -0,0 +1,166 @@
+                 bool useCoreSubgroups,
+                 std::vector<std::string> required_extensions = {})
+    {
+        device_ = device;
```
Surely these should all be as initialisers on the constructor?
And the std::vector<> should be a std::move() from the argument as well, to avoid copies?

polsenferreira commented on test_conformance/subgroups/test_workgroup_extended_types.cpp:
```diff
@@ -0,0 +1,166 @@
+    run_for_type(cl_device_id device, cl_context context,
+                 cl_command_queue queue, int num_elements,
+                 bool useCoreSubgroups,
+                 std::vector<std::string> required_extensions = {})
```
Looking at where this object will ultimately be constructed, it seems this should be a vector of const char * to avoid the std::string construction.
Alternative, use std::string_view but that requires upping the C++ standard requirements to 17.

polsenferreira commented on test_conformance/subgroups/test_workgroup_extended_types.cpp:
```diff
@@ -0,0 +1,166 @@
+                                             int num_elements)
+{
+    int error;
+    std::vector<std::string> required_extensions;
```
These heap allocations don't seem necessary.

polsenferreira commented on test_conformance/subgroups/test_workgroup_extended_types.cpp:
```diff
@@ -0,0 +1,166 @@
+#include "workgroup_common_templates.h"
+#include "harness/typeWrappers.h"
+
+struct run_for_type
```
This should be in anon namespace if the type is local to this file.

polsenferreira commented on test_conformance/subgroups/workgroup_common_templates.h:
```diff
@@ -0,0 +1,585 @@
+
+
+typedef std::bitset<128> bs128;
+inline cl_uint4 generate_bit_mask(cl_uint subgroup_local_id,
```
You probably meant static instead of inline.

polsenferreira commented on test_conformance/subgroups/workgroup_common_templates.h:
```diff
@@ -0,0 +1,585 @@
+
+typedef std::bitset<128> bs128;
+inline cl_uint4 generate_bit_mask(cl_uint subgroup_local_id,
+                                  std::string mask_type,
```
Does this really need to be a copy? It seems to be that mask_type is not being modified.

polsenferreira commented on test_conformance/subgroups/workgroup_common_templates.h:
```diff
@@ -0,0 +1,585 @@
+// work_item Which = 2 -  sub_group_non_uniform_broadcast - same as type 0 but
+// only 4 work_items from subgroup enter the code (are active)
+
+template <typename Ty, int Which = 0> struct BC
```
If this type is local to the file, it probably should be enclosed in an anon namespace

gwawiork replied:
this is used in many files

polsenferreira commented on test_conformance/subgroups/workgroup_common_templates.h:
```diff
@@ -0,0 +1,585 @@
+// subgroup takes only one value from only one chosen (the smallest subgroup ID)
+// work_item Which = 2 -  sub_group_non_uniform_broadcast - same as type 0 but
+// only 4 work_items from subgroup enter the code (are active)
+
```
Extraneous blank line.

polsenferreira commented on test_conformance/subgroups/workgroup_common_templates.h:
```diff
@@ -0,0 +1,585 @@
+// work_item Which = 2 -  sub_group_non_uniform_broadcast - same as type 0 but
+// only 4 work_items from subgroup enter the code (are active)
+
+template <typename Ty, int Which = 0> struct BC
```
This "Which" should really be an enum.

polsenferreira commented on test_conformance/subgroups/workgroup_common_templates.h:
```diff
@@ -0,0 +1,585 @@
+    }
+};
+
+template <typename Ty, int Which> struct OPERATION;
```
All these templates should be condensed into a switch on "Which". Let the compiler constant-propagate and remove the unused cases.
Would make this far simpler to understand and shorter.

polsenferreira commented on test_conformance/subgroups/workgroup_common_templates.h:
```diff
@@ -0,0 +1,585 @@
+
+static float to_float(subgroups::cl_half x) { return cl_half_to_float(x.data); }
+
+static subgroups::cl_half to_half(float x)
```
These templates could probably be better suited for operator overloading on subgroups::cl_half. Would allow code sharing with other types that are able to use the regular operators.

gwawiork replied:
Can this stay as is ?

polsenferreira replied:
I won't block the merge myself. We can check if others have differing opinions.

polsenferreira commented on test_conformance/subgroups/workgroup_common_templates.h:
```diff
@@ -0,0 +1,585 @@
+};
+
+// Scan Exclusive functions
+template <typename Ty, int Which> struct SCEX
```
This Which is being used where a functor might be more appropriate/logical.

polsenferreira commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -22,6 +22,16 @@
+#undef max
+
+// Adjust these individually below if desired/needed
+#define G 2000
+#define L 200
+
+#define NON_UNIFORM 4
```
These should definitely be proper constants in a relevant namespace, correctly documented.

There is no indication as to what they mean.

gwawiork replied:
Is this ok ?
`#define NON_UNIFORM_WG_SIZE 4`

polsenferreira commented on test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp:
```diff
@@ -0,0 +1,344 @@
+    {
+        bool taa = false;
+        for (int i = 0; i < NON_UNIFORM; ++i)
+            taa |= mx[i]
+                != 0; // return non zero if value non zero at least for one
```
Breaking this expression in two lines is confusing.

Just below, on AAN_OPERANTION<Ty, 1> you wrote the expression in a single line, which is much more clear.

polsenferreira commented on test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp:
```diff
@@ -0,0 +1,344 @@
+{
+    static void gen(Ty *x, Ty *t, cl_int *m, int ns, int nw, int ng)
+    {
+        int i, ii, j, k, n;
```
These are C89 style declarations - fold these to their initialisation points, the for-loops. Restrict their scope.

gwawiork replied:
I think such change is not important now

polsenferreira commented on test_conformance/subgroups/test_workgroup_non_uniform_arithmetic.cpp:
```diff
@@ -0,0 +1,774 @@
+                   int ng)
+    {
+        int ii, i, j, k, n;
+        int nj = (nw + ns - 1) / ns;
```
What's "nj"? What's it supposed to represent?

gwawiork replied:
nj - number of subgroup

polsenferreira commented on test_conformance/subgroups/test_workgroup_non_uniform_arithmetic.cpp:
```diff
@@ -0,0 +1,774 @@
+        genrand<Ty, Which>(x, t, m, ns, nw, ng);
+    }
+
+    static int chk(Ty *x, Ty *y, Ty *mx, Ty *my, cl_int *m, int ns, int nw,
```
What are all these parameters?

gwawiork replied:
It was really confusing form me as well. Such code construction was from the beginning. I just used it. I've added comments inside of code to be able understand these names but I can write once again:

x - array with golden data calculated on host side
y - array with data read from the device (to compare)
mx - set of data extracted from x for given local workgoup
my - set of data extracted from y for given local workgoup
m - is an array which stores helpful information for test execution. It has values set by device (subgroup local id, subgroup id). Additionally it can contain different information set on host for subgroup_brodcast there is index of work_item which value will be broadcasted
nw - size of local work group
ng - number of local work groups
ns - size of subgroup
nj - number of subgroup

gwawiork replied:
I prefer adding such comment. Renaming variables should be as separate pull request.

gwawiork replied:
some improvements here in https://github.com/KhronosGroup/OpenCL-CTS/pull/1074/commits/d9fd232dd04e3ef1c479f33ad05204a329a0ab75  in code for ballot functions

polsenferreira commented on test_conformance/subgroups/test_workgroup_non_uniform_arithmetic.cpp:
```diff
@@ -0,0 +1,774 @@
+
+
+// Scan Exclusive non uniform functions
+// Which: 0 - add, 1 - max, 2 - min, 3 - mul, 4 - and, 5 - or, 6 - xor, 7 -
+// logical and, 8 - logical or, 9 - logical xor
```
This sort of documentation makes it very obvious that Which should be an enum.

polsenferreira commented on test_conformance/subgroups/test_workgroup_non_uniform_arithmetic.cpp:
```diff
@@ -0,0 +1,774 @@
+                                                     cl_command_queue queue,
+                                                     int num_elements)
+{
+    int error;
```
C89 style declaration. Move down to first initialisation.

gwawiork replied:
Fixed

gwawiork commented:
@polsenferreira thanks for the review I will try to address them ASAP.

gwawiork commented:
@polsenferreira I think all your remarks have been addressed. Please look now in a free time. Thanks

polsenferreira commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -27,11 +27,145 @@
+{
+    switch (operation)
+    {
+        case ArithmeticOp::add_: return "add"; break;
+        case ArithmeticOp::max_: return "max"; break;
+        case ArithmeticOp::min_: return "min"; break;
+        case ArithmeticOp::mul_: return "mul"; break;
+        case ArithmeticOp::and_: return "and"; break;
+        case ArithmeticOp::or_: return "or"; break;
+        case ArithmeticOp::xor_: return "xor"; break;
+        case ArithmeticOp::logical_and: return "logical_and"; break;
+        case ArithmeticOp::logical_or: return "logical_or"; break;
+        case ArithmeticOp::logical_xor: return "logical_xor"; break;
```
All the trailing "break;" on these lines are dead. Some compilers may issue warnings here.
This is repeated elsewhere on this file.

gwawiork replied:
yes they are not needed

polsenferreira commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -27,11 +27,145 @@
+        case ArithmeticOp::logical_and: return "logical_and"; break;
+        case ArithmeticOp::logical_or: return "logical_or"; break;
+        case ArithmeticOp::logical_xor: return "logical_xor"; break;
+        default: log_error("Unknown operation request"); break;
```
This case seems to be "impossible" or occur only on internal bugs. I think this ought to abort the process, but that's just my opinion - perhaps others can comment on this.

polsenferreira commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -190,17 +328,17 @@ template <typename Ty> struct CommonTypeManager
     typedef std::false_type is_sb_scalar_type;
     static const bool type_supported(cl_device_id) { return true; }
-    static const Ty identify_limits(unsigned int test_id)
+    static const Ty identify_limits(ArithmeticOp operation)
```
This is much _much_ better!

polsenferreira commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -1098,8 +1239,8 @@ struct test
-        kernel_sstr << "#define NON_UNIFORM " + std::to_string(NON_UNIFORM)
-                + " \n";
+        kernel_sstr << "#define NON_UNIFORM_WG_SIZE "
+                + std::to_string(NON_UNIFORM_WG_SIZE) + " \n";
```
I don't understand. kernel_sstr is a stringstream - shouldn't these "+"s be "<<" instead?
It seems to me you're generating temp strings (from the  '+') unnecessarily.

polsenferreira commented on test_conformance/subgroups/test_barrier.cpp:
```diff
@@ -147,13 +147,13 @@ int test_barrier_functions(cl_device_id device, cl_context context,
     // Adjust these individually below if desired/needed
-#define G 2000
-#define L 200
+#define GWS 2000
```
The naming is better, but it is still undocumented.
I assume by name, this is the Global Workgroup Size, in which case it should have a
// Global Work Group Size used in {...}
I'd also like a justification for this being a preprocessor macro instead of an actual constant:
constexpr uint32_t GlobalWorkSize = 2000;

polsenferreira commented on test_conformance/subgroups/test_workgroup.cpp:
```diff
@@ -143,45 +156,41 @@ struct run_for_type
             device_, context_, queue_, num_elements_, "test_redadd",
             redadd_source, 0, useCoreSubgroups_);
-        error |= test<T, RED<T, 1>, G, L>::run(
+        error |= test<T, RED<T, ArithmeticOp::max_>, GWS, LWS>::run(
```
These look much better now, thank you.

polsenferreira commented on test_conformance/subgroups/workgroup_common_templates.h:
```diff
@@ -19,11 +19,12 @@
-inline cl_uint4 generate_bit_mask(cl_uint subgroup_local_id,
-                                  std::string mask_type,
+static cl_uint4 generate_bit_mask(cl_uint subgroup_local_id,
+                                  const char *mask_type,
```
This change is correct - it aviods unecessary heap allocations.
However, the code below needs to be modified to build. You'll need to replace the equalities with std::strcmp(mask_type, "le") - this will restore the MacOS build (https://github.com/KhronosGroup/OpenCL-CTS/pull/1074/checks?check_run_id=1649944477)

gwawiork replied:
I've decided change to const std::string&

polsenferreira commented on test_conformance/subgroups/workgroup_common_templates.h:
```diff
@@ -293,54 +225,94 @@ static subgroups::cl_half to_half(float x)
-template <> struct OPERATION<subgroups::cl_half, 1>
+    return 0;
+}
+// speciallize for floating points.
```
Typo: "Specialize"

polsenferreira commented on test_conformance/subgroups/workgroup_common_templates.h:
```diff
@@ -293,54 +225,94 @@ static subgroups::cl_half to_half(float x)
+}
+// speciallize for floating points.
+template <>
+inline cl_double calculate(cl_double a, cl_double b, ArithmeticOp operation)
```
I'm not sure the inline attribute on templates makes any difference.

gwawiork replied:
I got compiler errors on Linux while Linking.

polsenferreira commented on test_conformance/subgroups/workgroup_common_kernels.cpp:
```diff
@@ -0,0 +1,104 @@
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    out[gid] = sub_group_scan_inclusive_add(in[gid]);\n"
```
I think this was brought up before.
Are there any provisions to ensure that implementations are using the correct overload?
That is, what's stopping an implementation from type-promoting "char" to "int" and using the "int" overload of this function?

gwawiork replied:
maybe such check should work?  `sizeof(int[gid]) == sizeof(sub_group_scan_inclusive_add(in[gid]))`

polsenferreira replied:
Hm, that might.

gwawiork replied:
Example  [commit](https://github.com/KhronosGroup/OpenCL-CTS/pull/1074/commits/65cd0ad1ea6f8adc3344417f3d04892e98b933f6) for cluster reduce function has been added. Test runs ok. If you are ok with that I can add such solution for other functions.

work_group_functions_clustered_reduce...
  sub_group_clustered_reduce_add(int, 4 bytes) ...
  sub_group_clustered_reduce_max(int, 4 bytes) ...
  sub_group_clustered_reduce_min(int, 4 bytes) ...
  sub_group_clustered_reduce_mul(int, 4 bytes) ...
  sub_group_clustered_reduce_and(int, 4 bytes) ...
  sub_group_clustered_reduce_or(int, 4 bytes) ...
  sub_group_clustered_reduce_xor(int, 4 bytes) ...
  sub_group_clustered_reduce_add(uint, 4 bytes) ...
  sub_group_clustered_reduce_max(uint, 4 bytes) ...
  sub_group_clustered_reduce_min(uint, 4 bytes) ...
  sub_group_clustered_reduce_mul(uint, 4 bytes) ...
  sub_group_clustered_reduce_and(uint, 4 bytes) ...
  sub_group_clustered_reduce_or(uint, 4 bytes) ...
  sub_group_clustered_reduce_xor(uint, 4 bytes) ...
  sub_group_clustered_reduce_add(long, 8 bytes) ...
  sub_group_clustered_reduce_max(long, 8 bytes) ...
  sub_group_clustered_reduce_min(long, 8 bytes) ...
  sub_group_clustered_reduce_mul(long, 8 bytes) ...
  sub_group_clustered_reduce_and(long, 8 bytes) ...
  sub_group_clustered_reduce_or(long, 8 bytes) ...
  sub_group_clustered_reduce_xor(long, 8 bytes) ...
  sub_group_clustered_reduce_add(ulong, 8 bytes) ...
  sub_group_clustered_reduce_max(ulong, 8 bytes) ...
  sub_group_clustered_reduce_min(ulong, 8 bytes) ...
  sub_group_clustered_reduce_mul(ulong, 8 bytes) ...
  sub_group_clustered_reduce_and(ulong, 8 bytes) ...
  sub_group_clustered_reduce_or(ulong, 8 bytes) ...
  sub_group_clustered_reduce_xor(ulong, 8 bytes) ...
  sub_group_clustered_reduce_add(short, 2 bytes) ...
  sub_group_clustered_reduce_max(short, 2 bytes) ...
  sub_group_clustered_reduce_min(short, 2 bytes) ...
  sub_group_clustered_reduce_mul(short, 2 bytes) ...
  sub_group_clustered_reduce_and(short, 2 bytes) ...
  sub_group_clustered_reduce_or(short, 2 bytes) ...
  sub_group_clustered_reduce_xor(short, 2 bytes) ...
  sub_group_clustered_reduce_add(ushort, 2 bytes) ...
  sub_group_clustered_reduce_max(ushort, 2 bytes) ...
  sub_group_clustered_reduce_min(ushort, 2 bytes) ...
  sub_group_clustered_reduce_mul(ushort, 2 bytes) ...
  sub_group_clustered_reduce_and(ushort, 2 bytes) ...
  sub_group_clustered_reduce_or(ushort, 2 bytes) ...
  sub_group_clustered_reduce_xor(ushort, 2 bytes) ...
  sub_group_clustered_reduce_add(uchar, 1 bytes) ...
  sub_group_clustered_reduce_max(uchar, 1 bytes) ...
  sub_group_clustered_reduce_min(uchar, 1 bytes) ...
  sub_group_clustered_reduce_mul(uchar, 1 bytes) ...
  sub_group_clustered_reduce_and(uchar, 1 bytes) ...
  sub_group_clustered_reduce_or(uchar, 1 bytes) ...
  sub_group_clustered_reduce_xor(uchar, 1 bytes) ...

mantognini commented on test_conformance/subgroups/test_workgroup_ballot.cpp:
```diff
@@ -0,0 +1,973 @@
+                    (mx[ii + i].s0 != 0) ? bs.set(i) : bs.reset(i);
+                }
+                // convert bs to uint4
+                auto const uint_mask = bs128{ static_cast<unsigned long>(-1) };
```
If on a platform `unsigned long` is not 128 bits, this might not work. It would probably be safer to use `std::bitset::flip` or the `~` operator. This could be encapsulated in a free function if it's needed in a few places and you want to keep the `const`.

(untested)
```cpp
bs128 get_uint_mask() { bs128 mask; mask.flip(); return mask; }
```

mantognini commented on test_conformance/subgroups/test_workgroup_ballot.cpp:
```diff
@@ -0,0 +1,973 @@
+                    // Initialize data matrix indexed by local id and sub group
+                    // id
+                    e = (int)(genrand_int32(gMTdata) % 3);
+                    switch (e)
```
This switch looks very similar to some I've seen above. It might be worth DRY-ing them and introduce a function.

gwawiork replied:
removed

gwawiork replied:
Fixed https://github.com/KhronosGroup/OpenCL-CTS/pull/1074/commits/0bf22751ded07d81b6ad948105046cac61a45476

mantognini commented on test_conformance/subgroups/test_workgroup_ballot.cpp:
```diff
@@ -0,0 +1,973 @@
+                else if (operation == BallotOp::ballot_find_lsb
+                         || operation == BallotOp::ballot_find_msb)
+                {
+                    // Regarding to the spec, find lsb and find msb result is
```
This comment is slightly incorrect I believe. The value needs at least one bit set for a work-item in the subgroup. If the subgroup is very small and no bit is set for these work-items, this would also be UB.

Maybe another way to generate valid data would be to:
 1. pick a random number `x` (the domain is not important),
 2. pick a random value `i` in the domain `[0, subgroup_size]`.
 3. set the `i`-th bit of `x` to `1`.
 4. set `t[ii]` to `x`.

What is your opinion?

gwawiork replied:
Spec says that in case of no bit set (zeros on all bits in value) then behaviour is undefined. I've simplified a little bit generation values. Now it calculates values for each work item in the subgroup:
```
char x = (genrand_int32(gMTdata)) & 0xff;
// undefined behaviour in case of 0;
x = x ? x : 1;
memset(&t[wg_offset + wi_id], x, sizeof(Ty));
```

mantognini reviewed (commented):
That's clearly a non-negligible amount of work! :tada:  I've focused on the elements related to the ballot extension for this review. I didn't look at everything tough.

With respect to coverage, I've noticed that divergent subgroup are not tested. The spec reads:

> These functions [for the cl_khr_subgroup_ballot extension] need not be encountered by all work items in a subgroup executing the kernel.

That could be translated to kernels with an `if` statement, see this rough pseudo-code for example:
```
kernel foo(...) {
  if (input[id].is_active) {
    output[id] = sub_group_broadcast_first(input[id].value);
  }
}
```
(Although that's not to say it should be done that exact way.)

I didn't look into the machinery so I wonder: are non-uniform work-groups being covered?

gwawiork commented on test_conformance/subgroups/test_workgroup_ballot.cpp:
```diff
@@ -0,0 +1,973 @@
+    " if (xy[gid].x < NON_UNIFORM_WG_SIZE) {\n" // broadcast 4 values , other
+                                                // values are
+                                                // 0
+    "    out[gid] = sub_group_broadcast(x, xy[gid].z);\n"
```
Fixed in commit https://github.com/KhronosGroup/OpenCL-CTS/pull/1074/commits/6c7ddfc932f244fa0191b620aca92ed4db8d0eef

gwawiork commented on test_conformance/subgroups/test_workgroup_ballot.cpp:
```diff
@@ -0,0 +1,973 @@
+//
+// Copyright (c) 2017 The Khronos Group Inc.
```
Fixed

gwawiork commented on test_conformance/subgroups/test_workgroup_ballot.cpp:
```diff
@@ -0,0 +1,973 @@
+};
+
+static const char *bcast_non_uniform_source =
+    "__kernel void test_bcast_non_uniform(const __global Type *in, __global "
```
Could this be done as new PR after this one will be merged ?

mantognini replied:
Yes, I don't see why not.

gwawiork commented on test_conformance/subgroups/test_workgroup_ballot.cpp:
```diff
@@ -0,0 +1,973 @@
+                n = ii + ns > nw ? nw - ii : ns;
+                e = (int)(genrand_int32(gMTdata) % 3);
+
+                // Initialize data matrix indexed by local id and sub group id
```
It was so long ago... :) I see there was a need to generate 2 data patters to have two kind values > 0 and == 0 . Then verification function - `chk` - uses this to reset or set bit for particular work item in the subgroup (such simulation sub_group_ballot fnction on host side).
```
                for (i = 0; i < n; ++i)
                { 
                    (mx[ii + i].s0 != 0) ? bs.set(i) : bs.reset(i);
                }
```
 After that these values are compared with values obtained from device

mantognini replied:
I see now that `t`, in `gen`, and `mx`, in `chk`, are the supposed to be the same if I read code in `subhelpers.h` correctly. Yet, both `gen` and `chk` write data to these variables. Is that expected? By reading the code, I cannot tell whether some values are overridden intentionally or not.

NB: I haven't checked other variables, so there may be similar "surprises".

mantognini replied:
> In the gen function t stores calculated values for local workgroup (all subgroups in the LWG). Next data for all local workgroups is stored in variable x.
In the chk again data for local work group are copied to mx for verification.
So the flow looks t->x->mx. I do not see here any variable is overridden.

So IIUC, `t` and `mx` could be local variables?

gwawiork replied:
In my opinion looks like yes, they could be local. I will mark this as issue to be fixed

gwawiork commented:
> That's clearly a non-negligible amount of work! 🎉 I've focused on the elements related to the ballot extension for this review. I didn't look at everything tough.
> 
> With respect to coverage, I've noticed that divergent subgroup are not tested. The spec reads:
> 
> > These functions [for the cl_khr_subgroup_ballot extension] need not be encountered by all work items in a subgroup executing the kernel.
> 
> That could be translated to kernels with an `if` statement, see this rough pseudo-code for example:
> 
> ```
> kernel foo(...) {
>   if (input[id].is_active) {
>     output[id] = sub_group_broadcast_first(input[id].value);
>   }
> }
> ```
> 
> (Although that's not to say it should be done that exact way.)
> 
> I didn't look into the machinery so I wonder: are non-uniform work-groups being covered?

Thanks for review. I'm sorry if my answers are not too detailed. This PR has so many line codes that even for me it is hard to remind some details. I am thinking about this sentence `active work items` and wondering if this is the same as non uniform workgroup ? I think in current version of code I have that all work items are active. It should be added. But again for me it would be better add this as separate PR because large amount of commits in on PR make confusion

mantognini reviewed (commented):
> I am thinking about this sentence active work items and wondering if this is the same as non uniform workgroup ?

Those are two orthogonal concepts: you can have inactive work-items regardless of non-uniform/uniform workgroup sizes.

> It should be added. But again for me it would be better add this as separate PR because large amount of commits in on PR make confusion

That's fine from my personal point of view. I suppose it should be captured in an issue with the other follow-up elements so we don't collectively forget about it though.

gwawiork commented:
In the `gen` function `t` stores calculated values for local workgroup (all subgroups in the LWG).  Next data for all local workgroups is stored in variable `x`. 
In the `chk` again data for local work group are copied to `mx` for verification. 
So the flow looks t->x->mx. I do not see here any variable is overridden.

StuartDBrady commented:
> Those are two orthogonal concepts: you can have inactive work-items regardless of non-uniform/uniform workgroup sizes.

It would appear that non-uniform workgroup sizes are not properly tested, either. As it seems plausible that this affected existing tests, it is not obvious to me whether this would be best addressed in this PR or by a followup PR.

gwawiork commented:
> > Those are two orthogonal concepts: you can have inactive work-items regardless of non-uniform/uniform workgroup sizes.
> 
> It would appear that non-uniform workgroup sizes are not properly tested, either. As it seems plausible that this affected existing tests, it is not obvious to me whether this would be best addressed in this PR or by a followup PR.

Probably yes. Am I right that  non uniform should be set by adjusting LWS to GWS for eg. LWS = 5 GWS = 24 ?
Regarding active -  non active workitems, could  somebody describe situations when workitem can be inactive in the subgroups ? I cannot find any description in documentation.

StuartDBrady commented:
> Probably yes. Am I right that non uniform should be set by adjusting LWS to GWS for eg. LWS = 5 GWS = 24 ?

I believe so. "Any single dimension for which the global size is not divisible by the local size will be partitioned into two regions" according to the OpenCL specification. Arguably, this should be tested with multiple dimensions. It should be possible to test this for the 3D case with seven remainder regions (out of 8 regions in total).

On reflection, I realise that what I _really_ meant is that we don't thoroughly test the case where a subgroup's actual size is not equal to the maximum subgroup size. This occurs when the number of workitems within a workgroup is not the exact product of the maximum subgroup size and the number of subgroups. It would seem to me that this can still occur when using uniform workgroup sizes, but it is possible that I am mistaken.

To put it another way, this may occur in the case where the local size in a given dimension (or products thereof in multiple dimensions) is not an exact multiple of the maximum subgroup size. As the OpenCL specification also states "the mapping of work-items to sub-groups is implementation-defined", it is possible that there might not be a single methodology that would suit all possible implementations. However, I expect that it would be possible to test in a way that would suit most realistic implementations.

> Regarding active - non active workitems, could somebody describe situations when workitem can be inactive in the subgroups ? I cannot find any description in documentation.

That's because it's not a runtime API concept.

Consider the following example:
```
kernel void divergence_test(global int *result) {
    uint4 full_ballot = sub_group_ballot(1);
    uint divergence_mask;
    uint4 partial_ballot;

    if (get_sub_group_local_id() & 1) {
        divergence_mask = 0xaaaaaaaa;
        partial_ballot = sub_group_ballot(1);
    } else {
        divergence_mask = 0x55555555;
        partial_ballot = sub_group_ballot(1);
    }

    uint4 masked_ballot = full_ballot;
    masked_ballot.x &= divergence_mask;
    masked_ballot.y &= divergence_mask;
    masked_ballot.z &= divergence_mask;
    masked_ballot.w &= divergence_mask;

    result[get_global_id(0)] = all(masked_ballot == partial_ballot);
}
```

Here, we would expect `result` to be filled with the value `1` after execution, because we expect `masked_ballot` and `partial_ballot` to be equal for each workitem.

(Disclaimer: I have not tested the above, but it should hopefully serve to demonstrate the concept. Note, however: [LLVM bug 46199](https://bugs.llvm.org/show_bug.cgi?id=46199) which may also be of some interest.)

AnastasiaStulova commented:
We have recently added a new folder for extensions. @kpet, I guess the idea is that all extensions should be added there? 

But I suggest doing this in a separate change. Perhaps it should be captured in the same separate issue that @mantognini mentioned.

kpet commented:
> We have recently added a new folder for extensions. @kpet, I guess the idea is that all extensions should be added there?

For non-KHR extensions, definitely! For KHR extensions we're yet to have a discussion about this. As you say, we should not consider this under this PR.

AnastasiaStulova commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -354,25 +1231,55 @@ struct test
+                         extension.c_str(), kname, TypeManager<Ty>::name());
+                return 0;
+            }
+            kernel_sstr << "#pragma OPENCL EXTENSION " + extension
```
The extended subgroups don't require the pragmas, so I think this line should be removed!

AnastasiaStulova replied:
FYI I don't think we need the pragmas for cl_khr_subgroups either for the functionality being tested by the PR.

mantognini replied:
I agree, these should not be necessary as the spec doesn't say they are necessary. I suppose that this change can be done later, once https://github.com/KhronosGroup/OpenCL-Docs/issues/82 is resolved for example.

Aside from that, there are currently three ways to introduce `#pragma` in the kernel used by these tests:
1. When `doublen` or `halfn` are used.
2. When `useCoreSubgroups` is true.
3. When `required_extensions` is non-empty.

In other words, this cleanup would simplify the code in this function as only 1. would be necessary.

gwawiork commented:
> > Probably yes. Am I right that non uniform should be set by adjusting LWS to GWS for eg. LWS = 5 GWS = 24 ?
> 
> I believe so. "Any single dimension for which the global size is not divisible by the local size will be partitioned into two regions" according to the OpenCL specification. Arguably, this should be tested with multiple dimensions. It should be possible to test this for the 3D case with seven remainder regions (out of 8 regions in total).
> 
> On reflection, I realise that what I _really_ meant is that we don't thoroughly test cases where a subgroup's actual size is not equal to the maximum subgroup size. This occurs when the number of workitems within a workgroup is not the exact product of the maximum subgroup size and the number of subgroups in the workgroup. It would seem to me that this can still occur when using uniform workgroup sizes, but it is possible that I am mistaken.
> 
> To put it another way, this may occur in the case where the local size in a given dimension (or products thereof in multiple dimensions) is not an exact multiple of the maximum subgroup size. As the OpenCL specification also states "the mapping of work-items to sub-groups is implementation-defined", it is possible that there might not be a single methodology that would suit all possible implementations. However, I expect that it would be possible to test in a way that would suit most realistic implementations.
> 
> > Regarding active - non active workitems, could somebody describe situations when workitem can be inactive in the subgroups ? I cannot find any description in documentation.
> 
> That's because _non-uniform control-flow_ is not a runtime API concept (or even an OpenCL C language concept, it would seem).
> 
> Consider the following example:
> 
> ```
> kernel void divergence_test(global int *result) {
>     uint4 full_ballot = sub_group_ballot(1);
>     uint divergence_mask;
>     uint4 partial_ballot;
> 
>     if (get_sub_group_local_id() & 1) {
>         divergence_mask = 0xaaaaaaaa;
>         partial_ballot = sub_group_ballot(1);
>     } else {
>         divergence_mask = 0x55555555;
>         partial_ballot = sub_group_ballot(1);
>     }
> 
>     uint4 masked_ballot = full_ballot;
>     masked_ballot.x &= divergence_mask;
>     masked_ballot.y &= divergence_mask;
>     masked_ballot.z &= divergence_mask;
>     masked_ballot.w &= divergence_mask;
> 
>     result[get_global_id(0)] = all(masked_ballot == partial_ballot);
> }
> ```
> 
> Here, we would expect `result` to be filled with the value `1` after execution, because we expect `masked_ballot` and `partial_ballot` to be equal for each workitem.
> 
> (Disclaimer: I have not tested the above, but it should hopefully serve to demonstrate the concept. Note, however: [LLVM bug 46199](https://bugs.llvm.org/show_bug.cgi?id=46199) which may also be of some interest.)

Thanks @StuartDBrady  for the example.

mantognini reviewed (commented):
Previously I looked mostly at the tests for ballots. I had a look at other bits of this PR and will try to cover other areas in the coming days.

Most comments in this review don't have to be carried out in this PR: some comments are about issues existing before this PR or about issues present in other parts of the CTS as well. I do also note that there are significant improvements in `subhelpers.h` around the `TypeManager` and the new enumerations. This really helps! I suspect the new `subgroup::cl_*` structs could also be very useful in other parts of the CTS to reduce code duplication.

I do wonder however if the PR should be split into multiple PRs? I realise it's quite late to ask for this though. The main reason I'm putting the idea forward is that parts of the PR could be merged sooner and simplify/speed up a bit the overall review.

gwawiork commented:
> > Probably yes. Am I right that non uniform should be set by adjusting LWS to GWS for eg. LWS = 5 GWS = 24 ?
> 
> I believe so. "Any single dimension for which the global size is not divisible by the local size will be partitioned into two regions" according to the OpenCL specification. Arguably, this should be tested with multiple dimensions. It should be possible to test this for the 3D case with seven remainder regions (out of 8 regions in total).
> 
> On reflection, I realise that what I _really_ meant is that we don't thoroughly test cases where a subgroup's actual size is not equal to the maximum subgroup size. This occurs when the number of workitems within a workgroup is not the exact product of the maximum subgroup size and the number of subgroups in the workgroup. It would seem to me that this can still occur when using uniform workgroup sizes, but it is possible that I am mistaken.
> 
> To put it another way, this may occur in the case where the local size in a given dimension (or products thereof in multiple dimensions) is not an exact multiple of the maximum subgroup size. As the OpenCL specification also states "the mapping of work-items to sub-groups is implementation-defined", it is possible that there might not be a single methodology that would suit all possible implementations. However, I expect that it would be possible to test in a way that would suit most realistic implementations.
> 
> > Regarding active - non active workitems, could somebody describe situations when workitem can be inactive in the subgroups ? I cannot find any description in documentation.
> 
> That's because _non-uniform control-flow_ is not a runtime API concept (or even an OpenCL C language concept, it would seem).
> 
> Consider the following example:
> 
> ```
> kernel void divergence_test(global int *result) {
>     uint4 full_ballot = sub_group_ballot(1);
>     uint divergence_mask;
>     uint4 partial_ballot;
> 
>     if (get_sub_group_local_id() & 1) {
>         divergence_mask = 0xaaaaaaaa;
>         partial_ballot = sub_group_ballot(1);
>     } else {
>         divergence_mask = 0x55555555;
>         partial_ballot = sub_group_ballot(1);
>     }
> 
>     uint4 masked_ballot = full_ballot;
>     masked_ballot.x &= divergence_mask;
>     masked_ballot.y &= divergence_mask;
>     masked_ballot.z &= divergence_mask;
>     masked_ballot.w &= divergence_mask;
> 
>     result[get_global_id(0)] = all(masked_ballot == partial_ballot);
> }
> ```
> 
> Here, we would expect `result` to be filled with the value `1` after execution, because we expect `masked_ballot` and `partial_ballot` to be equal for each workitem.
> 
> (Disclaimer: I have not tested the above, but it should hopefully serve to demonstrate the concept. Note, however: [LLVM bug 46199](https://bugs.llvm.org/show_bug.cgi?id=46199) which may also be of some interest.)

@StuartDBrady In latest changes I've added ballots improvements. To sum up:
- we test only 1D work groups (In my opinion it should be separate Issue to solve lack of more dimensions)
- I've added for some tests testing non uniform workgroup sizes
- I've modified kernels code in ballots to be more precisely in testing active/inactive work items (non uniform control flow)

gwawiork commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -19,9 +19,151 @@
 #include <vector>
+#include <type_traits>
+
+#undef min
```
I use `std::is_scalar` for scalar types detection

gwawiork commented on test_conformance/subgroups/procs.h:
```diff
@@ -56,5 +56,31 @@ extern int test_ifp_ext(cl_device_id device, cl_context context,
+                                                    cl_context context,
+                                                    cl_command_queue queue,
+                                                    int num_elements);
+extern int test_work_group_functions_non_uniform_vote(cl_device_id device,
```
ok noticed

gwawiork commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -19,9 +19,151 @@
+#undef min
+#undef max
+
+#define NON_UNIFORM_WG_SIZE 4
```
It has been changed to NR_OF_ACTIVE_WORK_ITEMS. It indicates how many work items are "active" in the control flow (if else statement). Why 4 probably: to be below possible subgroup size.

gwawiork commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -56,219 +198,949 @@ class subgroupsAPI {
 };
+};
+
+static bool int64_ok(cl_device_id device)
```
noticed. I will open new Issue

gwawiork commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -56,219 +198,949 @@ class subgroupsAPI {
+inline bool compare_ordered(const subgroups::cl_half &lhs,
+                            const subgroups::cl_half &rhs)
+{
+    return cl_half_to_float(lhs.data) == (rhs.data) && !is_half_nan(lhs.data);
```
Yes, thanks.

jlewis-austin replied:
I don't think the `!is_half_nan(lhs.data)` does anything here, because the first comparison succeeding implies no NaNs were involved. To exclude NaN comparisons (ie treat any two NaNs as equal regardless of mantissa contents), it would have to look something like 
`(cl_half_to_float(lhs.data) == cl_half_to_float(rhs.data)) || (is_half_nan(lhs.data) && is_half_nan(rhs.data))`

gwawiork commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -56,219 +198,949 @@ class subgroupsAPI {
+    static const char *name() { return ""; }
+    static const char *add_typedef() { return "\n"; }
+    typedef std::false_type is_vector_type;
+    typedef std::false_type is_sb_vector_size3;
+    typedef std::false_type is_sb_vector_type;
+    typedef std::false_type is_sb_scalar_type;
```
The data types from namespace subgroups

gwawiork commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -56,219 +198,949 @@ class subgroupsAPI {
+    }
 };
-template <> struct TypeIdentity<cl_uint, 2>
+template <typename T> struct TypeManager : public CommonTypeManager<T>
```
Ok fixed

gwawiork commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -56,219 +198,949 @@ class subgroupsAPI {
-    static const char *val() { return "typedef double Type;\n"; }
+    ::cl_half16 data;
 };
+};
```
ok fixed

mantognini commented on test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp:
```diff
@@ -0,0 +1,471 @@
+void calculate(const T *mx, NonUniformVoteOp operation, int subgroup_size,
+               bool *compare_result)
+{
+    compare_result[0] = 1;
+    compare_result[1] = 1;
```
Did you mean to use `true` instead of `1`? Or should the function produce `cl_int`? (There are other occurrences of this.)

mantognini reviewed (commented):
Thanks for the update and opening the follow-up issues. I haven't managed to go through all of the new content yet, but here are some additional comments. Again, not everything has to be done right away -- if you agree with my remarks, follow up tasks can be created.

mantognini commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -354,25 +1233,55 @@ struct test
+                log_info("The extension %s not supported on this device. SKIP "
+                         "testing - kernel %s data type %s\n",
+                         extension.c_str(), kname, TypeManager<Ty>::name());
+                return TEST_PASS;
```
Shouldn't that be `TEST_SKIP`?

mantognini commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -354,25 +1233,55 @@ struct test
+        if (!TypeManager<Ty>::type_supported(device))
+        {
+            log_info("Data type not supported : %s\n", TypeManager<Ty>::name());
+            return 0;
```
Shouldn't that be `TEST_SKIP`?

StuartDBrady commented on test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp:
```diff
@@ -0,0 +1,471 @@
+                { // for each work_item in subgroup
+                  // sum of output values should be 1
+
+                    my[ii + i] > 0 ? rr += 1 : rr += 0;
```
my[] contains results from the kernel i.e. values returned by `sub_group_elect()`, does it not?

The specification states "true (nonzero)" and "false (nonzero)" are the values that may be returned.  Why this this checking whether the result is _greater_ than 0?

StuartDBrady replied:
This seems to be _counting_ the number of elected work items in the subgroup, rather than checking that the correct item has been elected.

The work item that is elected must be the active work item with the smallest subgroup local ID, not just any arbitrary work item in the subgroup.

The CTS should be testing for this.

It is possible that I have misread something: I find the variable names used almost impenetrable.

StuartDBrady commented on test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp:
```diff
@@ -0,0 +1,471 @@
+                    my[ii + i] > 0 ? rr += 1 : rr += 0;
+                }
+                // expectation is that only one elected returned true
+                tr = n <= NR_OF_ACTIVE_WORK_ITEMS ? 1 : 2;
```
If we're expecting only one elected work item in the subgroup, why "2" here?

gwawiork replied:
fixed

StuartDBrady commented on test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp:
```diff
@@ -0,0 +1,471 @@
+    // one in subgroup true others false.
+    "        out[gid] = sub_group_elect();\n"
+    "    } else {\n"
+    "        out[gid] = sub_group_elect();\n"
```
If work items for which `xy[gid].x < NR_OF_ACTIVE_WORK_ITEMS` is true are all active, and work items for which `xy[gid].x < NR_OF_ACTIVE_WORK_ITEMS` is false are not active, then surely no threads will ever execute the `else` case here.

I presume that this is not the case and that `NR_OF_ACTIVE_WORK_ITEMS` is the number of work items for which the `if` condition should be true, and for the remaining items the `else` case is taken. If this is the case, the naming of `NR_OF_ACTIVE_WORK_ITEMS` is _really_ confusing and should be changed.

StuartDBrady replied:
Given how `xy[gid].x` is initialised, this always elects the work item with subgroup local ID 0 in every subgroup.

To ensure adequate coverage, work items that do not have a subgroup local ID of 0 should be elected for some subgroups as part of this testing.

Preferably, this would test that work items with each possible subgroup local ID are electable on implementations having a maximum subgroup size of 128. The test must still work properly on implementations with a smaller maximum subgroup size.

It is important that this at least test that work items with the maximum subgroup local ID are electable, i.e. that with a maximum subgroup size of 128, if no other work item is active, when the work item with local ID 127 calls `sub_group_elect()`, 1 is returned.

It is important that the test ensures that work items with a subgroup local ID between 64 and 95 are electable, for implementations where this is applicable.

It is important that the test ensures that work items with a subgroup local ID between 32 and 63 are electable, for implementations where this is applicable.

StuartDBrady replied:
Although the testing here is an improvement on what preceded it, there are still substantial gaps in coverage. I realise while most of what I said here should be true, the comment that local ID 0 is tested in every subgroup only applies to the `if` block, not the `else` block.

Until [LLVM bug 46199](https://bugs.llvm.org/show_bug.cgi?id=46199) is fixed it is unhelpful not to have a mode in which convergence is _not_ tested.

In both the `if` and `else` blocks, the active subgroups as would be reported by `sub_group_ballot(1)` are a contiguous range. It is important that this test use non-contiguous ranges of active subgroup local IDs.

Supposing an implementation was written to elect _any_ work item for which the work item with the preceding subgroup local ID was inactive or for which the preceding subgroup local ID would be -1. (E.g. if using a ballot to implement the election, and given `sub_group_ballot(1).x == 0xff00ff00`, suppose the implementation elects _two_ work item, one with subgroup local ID 8 and the other with subgroup local ID 24.) Any such implementation would be buggy but would still pass this test.

Supposing also an implementation was written to elect the work item with the least significant subgroup local ID out of the contiguous group of work items with the most significant local IDs in the subgroup. (E.g. if using a ballot to implement the election, and given `sub_group_ballot(1).x == 0xff00ff00`, suppose the implementation elects _one_ work item, with subgroup local ID 24.) Any such implementation would be buggy but would still pass this test.

gwawiork replied:
Then should I prepare test which run `n = subgroup_size` times kernel with parameter `elect_item` changing from `0` to `subgroup_size - 1` ?
```
if (xy[gid].x == elect_item)
sub_group_elect();
```

StuartDBrady replied:
That's not quite what I'm saying. It is important to test with a variety of sets of _active_ but unelected work items in the test.

The sorts of bit pattern used in memory testers might give you inspiration, here, e.g. 0x00ffff00, 0xff00ff00, 0xff0000ff, 0xff00ff00, 0x0f0f0f0f, 0x0f0ff0f0, 0xaaaa5555, 0x5555aaaa, 0x55aaaa55, some so on and so forth.

This doesn't have to be quite as elaborate at that, though, it just needs to cover relevant cases. Others may be able to better prescribe the testing needed. If you put forward a proposal I would be happy to review it, but I cannot see how there would be a textbook answer as to how to test this.

gwawiork replied:
Please verify this proposal: https://github.com/KhronosGroup/OpenCL-CTS/pull/1074/commits/65c43f3375942f0c923df9501912efc88081dfce

StuartDBrady commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -452,25 +1361,57 @@ struct test
     }
 };
 
+static void set_last_worgroup_params(int non_uniform_size,
```
worgroup -> workgroup

gwawiork replied:
fixed

StuartDBrady commented on test_conformance/subgroups/CMakeLists.txt:
```diff
@@ -7,6 +7,14 @@ set(${MODULE_NAME}_SOURCES
     test_ifp.cpp
+    test_workgroup_extended_types.cpp
+    workgroup_common_kernels.cpp
+    test_workgroup_non_uniform_vote.cpp
+    test_workgroup_non_uniform_arithmetic.cpp
+    test_workgroup_ballot.cpp
+    test_workgroup_clustered_reduce.cpp
+    test_workgroup_shuffle.cpp
+    test_workgroup_shuffle_relative.cpp
```
I find many of the file names (and the corresponding test names) here confusing. There are subgroup functions, not workgroup functions. For some of the existing tests perhaps it made sense as there were equivalent workgroup functions, but for the extended subgroups builtins, this is not the case.

gwawiork replied:
fixed

mantognini commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -354,25 +1233,55 @@ struct test
         clKernelWrapper kernel;
         cl_platform_id platform;
-        cl_int sgmap[2 * GSIZE];
+        cl_int sgmap[4 * GSIZE];
```
I believe some kernels where not updated to use `cl_int4` instead of `cl_int2`. For example, in `test_conformance/subgroups/test_barrier.cpp`. These kernels should also use the `XY` macro for consistency and decrease maintenance complexity.

(Github doesn't allow me to comment on lines that weren't changed by the PR...)

mantognini replied:
This variable and the related parameters could (and probably should) be replaced with
```cpp
cl_int4 sgmap[GSIZE];
```
as this would make the code easier to reason about and correspond more explicitly to the `xy` kernel parameters.

This can be done later, too, if desired.

mantognini commented on test_conformance/subgroups/workgroup_common_templates.h:
```diff
@@ -0,0 +1,778 @@
+
+            for (int i = 0; i < n; ++i)
+            {
+                cl_ulong x;
+                if ((operation == ArithmeticOp::add_
+                     || operation == ArithmeticOp::mul_)
+                    && is_floating_point<Ty>())
+                {
+                    // work around different results depending on operation
+                    // order by having input with little precision
+                    x = genrand_int32(gMTdata) % 64;
+                }
+                else
+                {
+                    x = genrand_int64(gMTdata);
+                    if ((operation == ArithmeticOp::logical_and
+                         || operation == ArithmeticOp::logical_or
+                         || operation == ArithmeticOp::logical_xor)
+                        && ((x >> 32) & 1) == 0)
+                        x = 0; // increase probability of false
+                }
+                set_value(t[ii + i], x);
```
The data generated might be problematic in some cases. I specifically looked at some of the existing tests, such as for `sub_group_scan_(ex|in)clusive_(add|min|max)` and `sub_group_reduce_(add|min|max)`. There might be issues with other functions, too.

1. For the min/max functions with floating point types, the specification doesn't say much about NaN: i.e. are they propagated like with `fmax` or is the result undefined as with `max`? I suppose we should consider the second, weaker case given the lack of precision in the spec. With that in mind, the generated data should not included NaNs or we wouldn't be able to validate the output.
2. For the add functions on signed integer types, the generated values can be arbitrarily large and cause overflows and therefore undefined behaviour (on both the device and the host to compute the reference values). Previously, the generated values where small:
   - random values in `[1, subgroup_size]` for reduce or scan exclusive functions;
   - the sequence `0 1 2 ... (subgroup_size - 1)` for scan inclusive functions.

There might be other issues.

While the new random data is not perfect, it is arguably better than the previous one. It's also very nice to have this function instead of duplicated code here and there.

I'd argue this should be dealt with a follow-up task, but we should be aware of these issues, especially regarding undefined behaviour/results.

mantognini commented on test_conformance/subgroups/workgroup_common_templates.h:
```diff
@@ -0,0 +1,778 @@
+    {
+        case ArithmeticOp::add_: return to_half(to_float(a) + to_float(b));
+        case ArithmeticOp::max_:
+            return to_float(a) > to_float(b) || is_half_nan(b.data) ? a : b;
+        case ArithmeticOp::min_:
+            return to_float(a) < to_float(b) || is_half_nan(b.data) ? a : b;
```
I'm not sure we can have NaN as inputs. See my comment below. I'm not sure whether this function should be impacted by this though.

gwawiork commented on test_conformance/subgroups/test_workgroup.cpp:
```diff
@@ -679,28 +204,26 @@ struct run_for_type
-
+}
 // Entry point from main
 int test_work_group_functions(cl_device_id device, cl_context context,
```
I do not understand. Now they are grouped in one place . This `test_work_group_functions` tests all (broadcast, scan inclusive, scan exclusive, reduce) functions in one place

mantognini replied:
In `test_work_group_functions_extended_types()` the relevant vector cases are being tested. Here only the scalar cases are covered.

I guess this is fine -- it's one valid design that group tests by extensions. Another design would be to group tests per function and enable/skip some subtests when an extension is not supported.Bboth these designs have their own pros & cons. My comment was more intended as an observation than a request for change. I support we can revisit this discussion at a later time if there is an interest.

Feel free to mark this thread as resolved if you feel it has little value on the short term.

gwawiork commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -452,25 +1361,57 @@ struct test
+
+template <typename Ty>
+static void set_randomdata_for_subgroup(Ty *workgroup, int wg_offset,
+                                        int current_sbs)
```
some shortcuts sometimes are helpful but I can add whole words to variables names if you preferer.

mantognini commented on test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp:
```diff
@@ -0,0 +1,504 @@
+    }
+}
+template <>
+void calculate(const int *mx, NonUniformVoteOp operation, int subgroup_size,
```
Why do we need a template specialisation here? Can't all the code be in the above function? It seems this copy/paste is not necessary.

mantognini commented on test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp:
```diff
@@ -0,0 +1,504 @@
+{
+    compare_result[0] = false;
+    compare_result[1] = false;
+    switch (operation)
```
Most of these could be shorter by using STL algorithms. (std::all_of, any_of, equal + small lambda)

StuartDBrady commented on test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp:
```diff
@@ -0,0 +1,504 @@
+
+                for (i = 0; i < n; ++i)
+                {
+                    i == remember_lowest_id ? tr = 1 : tr = 0;
```
Generally speaking, it's probably better from a debugging perspective if the kernel writes the exact value returned from the builtin, and normalization is done on the host side.

StuartDBrady commented on test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp:
```diff
@@ -0,0 +1,504 @@
+    error |= rft.run_nu_all_eq<cl_double>();
+    error |= rft.run_nu_all_eq<subgroups::cl_half>();
+    error |= rft.run_elect<int, 0xffffaaaa>();
+    error |= rft.run_elect<int, 0x00000000>();
```
The mask of 0 here is pointless, as the kernel will not call `sub_group_elect()` at all, resulting in unnecessary noise in the log output.

StuartDBrady replied:
It _would_ make sense to test with all work items active, but as has been said, there can be more than 32 work items in a subgroup.

gwawiork replied:
I think kernel is prepared in such way that each workgroup size will be properly handled.

gwawiork replied:
Oh, I will have to make a correction according to comment https://github.com/KhronosGroup/OpenCL-CTS/pull/1074#discussion_r576880173

StuartDBrady replied:
If you're assuming GitHub's UI gives any indication as to the exact comment you've replied to, please note that it does not - please quote comments if (and only if) there is ambiguity (or some other lack of clarity) as to what you're replying to.

> I think kernel is prepared in such way that each subgroup size will be properly handled.

I assume this in response to my question about the mask of 0x00000000.  Are you working under the false impression that this works in the way that a mask of 0xffffffff would actually work?  (Or am I perhaps mistaken about this?)

gwawiork replied:
> that this works in the way that a mask of 0xffffffff would actually work? (Or am I perhaps mistaken about this?)

I mean that fixing existing kernel it will work for any subgroup size, then testing mask 0xffffffff will work correctly

mantognini commented on test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp:
```diff
@@ -212,129 +125,70 @@ template <typename Ty, NonUniformVoteOp operation> struct AAN
-// Test for elect function.
-// Discover only one elected work item in subgroup - with the
-// lowest subgroup local id
-template <unsigned int work_items_mask> struct ELECT
```
Nice code folding.

mantognini commented on test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp:
```diff
@@ -407,48 +256,60 @@ struct run_for_type
+                                                        non_uniform_all_source,
+                                                        0, useCoreSubgroups_,
+                                                        required_extensions_,
+                                                        static_cast<int>(
```
BTW, these `run_*` functions look very similar. I see the mask is now a function parameter If you want to reduce code duplication further and avoid such copy/paste issues, you could create a base function, possibly with the following signature or something close to it:

```c++
template <typename T, unsigned int work_items_mask, NonUniformVoteOp op>
int run_impl(char const* kernel_name, char const* source)
```

It might also be possible to make `work_items_mask` a non-template argument by slightly modifying `gen`/`chk` functions and making `test::run` have extra variadic template arguments that are passed to `gen`/`chk`. This would reduce the number of template instantiation in the code. Just a thought.

gwawiork replied:
fixed

mantognini commented on test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp:
```diff
@@ -474,31 +335,173 @@ int test_work_group_functions_non_uniform_vote(cl_device_id device,
     run_for_type rft(device, context, queue, num_elements, true,
                      required_extensions);
 
-    int error = rft.run_nu_all_eq<cl_char>();
```
There is a lot of code duplication below. If the mask is no longer a template argument, a for loop could help here. Otherwise, I would recommend having function(s) to remove the code duplication. For example,

```c++
template <unsigned mask> int run_nu_all_eq(run_for_type rft) {
  int error = rft.run_nu_all_eq<cl_char, mask>();
  error |= rft.run_nu_all_eq<cl_uchar, mask>();
  ...
  return error;
}
```

There are other alternatives to shorten this function, I suppose.

gwawiork replied:
WorkGroupParams struct solve this issue

mantognini reviewed (commented):
I've marked a few discussion as resolved thanks to your update.

lakshmih commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -452,25 +1373,57 @@ struct test
         memset(&idata[0], 0, input_array_size * sizeof(Ty));
```
Shouldn't sgmap be memset here as well, since it gets accessed in the kernel for computing get_sub_group_local_id?

jlewis-austin commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -56,220 +199,957 @@ class subgroupsAPI {
+        switch (operation)
+        {
+            case ArithmeticOp::add_: return { 0x0000 };
+            case ArithmeticOp::max_: return { 0xfc00 };
```
Comments on these values would be helpful to translate each into reader-friendly form

jlewis-austin commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -56,220 +199,957 @@ class subgroupsAPI {
+    for (auto i = 0; i < size; ++i)
     {
-        return is_extension_available(device, "cl_khr_fp16");
+        // log_info("Values index i = %d got %lu expected %lu\n", i, lhs.s[i],
```
Any reason to keep this around?

gwawiork replied:
removed

jlewis-austin commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -56,220 +199,957 @@ class subgroupsAPI {
+                            const subgroups::cl_half &rhs)
+{
+    return cl_half_to_float(lhs.data) == cl_half_to_float(rhs.data)
+        && !is_half_nan(lhs.data);
```
I commented on this earlier but it dropped off. If lhs.data is NaN, the equality comparison will be false per IEEE-754 rules and the following NaN check irrelevant.
If the intent is to follow those rules it's harmless but confusing. A comment would be helpful to state the intended behavior.

gwawiork replied:
removed

jlewis-austin commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -56,220 +199,957 @@ class subgroupsAPI {
+}
+
+template <typename Ty>
+inline bool compare_ordered(const subgroups::cl_half &lhs, const int &rhs)
```
It's a bit unusual to provide an int to float comparison, as opposed to making the caller reconcile the types. The equality comparison is done in float with an implicit lossy promotion of the int to float, so there will cases where this claims equality for differing integer and float values (ie when the integer has over 24 significant bits). If that's acceptable for the purposes of these tests, it should be noted; if not, some argument checking would need to be done to ensure that the int argument can be exactly converted to float.

jlewis-austin commented on test_conformance/subgroups/workgroup_common_templates.h:
```diff
@@ -0,0 +1,778 @@
+static subgroups::cl_half to_half(float x)
+{
+    subgroups::cl_half value;
+    value.data = cl_half_from_double(x, CL_HALF_RTE);
```
cl_half_from_float should work here?

gwawiork replied:
fixed

jlewis-austin commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -56,220 +199,957 @@ class subgroupsAPI {
+{
+    static const char *name() { return "float"; }
+    static const char *add_typedef() { return "typedef float Type;\n"; }
+    static cl_float identify_limits(ArithmeticOp operation)
```
I was really baffled by what these functions do until I found where the spec talks about the identity _I_ for various operations. What does "identify_limits" refer to? For me something like get_identity_value would be much more clear

jlewis-austin commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -56,220 +199,957 @@ class subgroupsAPI {
-template <typename Ty> struct TypeName;
-template <> struct TypeName<cl_half>
+// Need to defined custom type for vector size = 3 and half type. This is
+// because because of 3-component types are otherwise indistinguishable from the
```
extra "because" typo

gwawiork replied:
fixed

jlewis-austin commented on test_conformance/subgroups/workgroup_common_templates.h:
```diff
@@ -0,0 +1,778 @@
+                     || operation == ArithmeticOp::mul_)
+                    && is_floating_point<Ty>())
+                {
+                    // work around different results depending on operation
```
There's a lot to keep track of here as the range-reduced but possibly sign extended integer is assigned to a cl_ulong with possible underflow. Then the lower 16 bits are assigned to a cl_half via modulo arithmetic. If I have all that right, you would end up with a distribution consisting of only NaNs (all 1s in the top bits) and positive denormals (only lowest 7 bits set).  Both are problematic for these tests. Let me know if I've missed the boat on this, but it feels like the floats will have to be constructed very carefully to ensure that they only consist of low-precision normal floats of similar magnitude and ideally both positive and negative values.

jlewis-austin commented on test_conformance/subgroups/workgroup_common_templates.h:
```diff
@@ -0,0 +1,778 @@
+
+            for (int i = 0; i < n; ++i)
+            {
+                cl_ulong x;
```
I'd like to avoid re-using one of the function argument names here.

gwawiork replied:
fixed

lakshmih commented:
[Qualcomm] We pass the test with modifications suggested in these comments:
https://github.com/KhronosGroup/OpenCL-CTS/pull/1074/#discussion_r592488170
https://github.com/KhronosGroup/OpenCL-CTS/pull/1074/#discussion_r592028509
Please note that we don't support extended subgroup functions and therefore have no input on the tests targeting those.

AnastasiaStulova commented on test_conformance/subgroups/workgroup_common_kernels.cpp:
```diff
@@ -0,0 +1,104 @@
+                           "    int gid = get_global_id(0);\n"
+                           "    XY(xy,gid);\n"
+                           "    Type x = in[gid];\n"
+                           "    out[gid] = sub_group_broadcast(x, xy[gid].z);\n"
```
Is this now fixed by https://github.com/KhronosGroup/OpenCL-CTS/pull/1074/commits/6202fd8d7135320f9f303fb87f93bb13d724161e?

gwawiork replied:
In my opinion yes

gwawiork commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -343,7 +1226,9 @@ struct test
-                   const char *src, int dynscl, bool useCoreSubgroups)
+                   const char *src, int dynscl, bool useCoreSubgroups,
+                   std::vector<std::string> const &required_extensions = {},
+                   unsigned int work_items_mask = 0)
```
fixed

gwawiork commented on test_conformance/subgroups/subhelpers.h:
```diff
@@ -1145,6 +1145,11 @@ inline bool compare_ordered(const subgroups::cl_half &lhs,
         && !is_half_nan(lhs.data);
 }
 
+template <typename Ty>
+inline bool compare_ordered(const subgroups::cl_half &lhs, const int &rhs)
+{
+    return cl_half_to_float(lhs.data) == rhs;
+}
```
regarding #undef min
The issue described here https://stackoverflow.com/questions/1394132/macro-and-member-function-conflict
Decided to use additional parenthesis. The undef will be removed

gwawiork replied:
fixed

gwawiork commented on test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp:
```diff
@@ -0,0 +1,504 @@
+    int elect_work_item = 1 << (get_sub_group_local_id() % get_sub_group_size());
+        if (elect_work_item & WORK_ITEMS_MASK){
+        out[gid] = sub_group_elect();
+        }
```
fixed

gwawiork commented on test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp:
```diff
@@ -16,122 +16,36 @@
+struct VOTE
 {
-    static void gen(Ty *x, Ty *t, cl_int *m, int ns, int nw, int ng)
+    static void gen(T *x, T *t, cl_int *m, int ns, int nw, int ng)
```
fixed

gwawiork commented on test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp:
```diff
@@ -212,129 +125,70 @@ template <typename Ty, NonUniformVoteOp operation> struct AAN
-
-                    int elect_work_item = 1 << i % n;
-                    if (work_items_mask & elect_work_item)
+                    int check_work_item = 1 << i % 32;
```
changed to std::uint32_t and added parentheses

gwawiork commented on test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp:
```diff
@@ -16,122 +16,36 @@
-        log_info("  sub_group_non_uniform_%s(%s)...\n",
-                 operation_names(operation), TypeManager<Ty>::name());
+
+        log_info("  sub_group_%s... \n", operation_names(operation));
```
fixed

gwawiork commented:
Thank you all for contribution and review. For today that is all I have to pushed out. Other issues will be address as separate GitHub Issues.

StuartDBrady commented on test_conformance/subgroups/test_subgroup_non_uniform_vote.cpp:
```diff
@@ -0,0 +1,303 @@
+        "cl_khr_subgroup_non_uniform_vote"
+    };
+
+    std::vector<uint32_t> masks{ 0xffffffff, 0x55aaaa55, 0x5555aaaa, 0xaaaa5555,
```
The masks here are specific to a maximum subgroup size of 32. This is not appropriate. Please rewrite using a bitset of 128 bits, with one bit per work item for a maximum subgroup size of 128. Please use appropriate bit patterns to ensure reasonable coverage, preferably generated computationally.

The testing here lacks coverage of elect operations where the elected subgroup has a subgroup local ID greater than or equal to 32. The testing here does not ensure that any subgroup local ID (within the maximum subgroup size) is electable.

I have drawn special attention to elect operations above, because these are the operations for which the gap in coverage is most likely to hide an implementation bug. However, it is also not appropriate that the testing for sub_group_non_uniform_all, sub_group_non_uniform_all and sub_group_non_uniform_all_equal also uses masks specific to a maximum subgroup size of 32. Testing for these also does not cover cases where all work items with a subgroup local ID below 32 are inactive, and does not reasonably ensure that correct results are obtained when arbitrary sets of work items are active.

Note that for cl_khr_subgroup_non_uniform_vote, there is no direct reference to a maximum possible subgroup size of 128, but as other extensions in the family _do_ place a limit on the maximum subgroup size of 128 work items, it is reasonable to apply that same limit here.

gwawiork replied:
In my opinion this code is not limiting only to subgroup size == 32.
Please look at the lines where active workitems are chosen:
in verification function:
https://github.com/gwawiork/OpenCL-CTS/blob/extsub_2711_2020_all_changes/test_conformance/subgroups/test_subgroup_non_uniform_vote.cpp#L144
in kernel:
https://github.com/gwawiork/OpenCL-CTS/blob/extsub_2711_2020_all_changes/test_conformance/subgroups/test_subgroup_non_uniform_vote.cpp#L222
When subgroup_local_id is bigger than 32 operation `%` cause that calculated number will be always between 0 and 31 to match mask 32 bit size.

StuartDBrady replied:
I have read the code several times. The testing lacks coverage of elect operations where the elected subgroup has a subgroup local ID greater than or equal to 32.

If I am incorrect, please provide a single example of an instance in which the sub_group_elect() call returns a value of 32 or greater.

StuartDBrady commented on test_conformance/subgroups/test_subgroup_ballot.cpp:
```diff
@@ -0,0 +1,1089 @@
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    Type x = in[gid];\n"
+    "    if (xy[gid].x < NR_OF_ACTIVE_WORK_ITEMS) {\n"
```
Please rewrite the testing of non-uniformity here to match that for cl_khr_subgroup_non_uniform_vote with changes to reasonably ensure correctness.

This comment applies to all kernels that test "< NR_OF_ACTIVE_WORK_ITEMS". This is not an adequate level of testing for any of the extended subgroup functions.

gwawiork replied:
This issue is addressed by #1198

StuartDBrady replied:
Agreed.

StuartDBrady commented on test_conformance/subgroups/test_subgroup_ballot.cpp:
```diff
@@ -0,0 +1,1089 @@
+    "uint4 partial_ballot;\n"
+    "uint gid = get_global_id(0);"
+    "XY(xy,gid);\n"
+    "if (get_sub_group_local_id() & 1) {\n"
```
It seems that the example I gave of divergence has been directly copied and pasting into the testing, here, without appropriate modifications.

The testing of non-uniform control flow should be similar (or identical) to sub_group_non_uniform_broadcast(), sub_group_elect() and in fact all of the extended subgroup functions that are sensitive to divergence of control flow within the subgroup.  Please update it to match, once the testing for sub_group_elect(), etc. correctly supports a maximum subgroup size of 128.

StuartDBrady replied:
This test is not a test of the sub_group_ballot() function.

It will pass even if the implementation of sub_group_ballot() unconditionally returns 0, always.

StuartDBrady replied:
Agreed.

StuartDBrady commented on test_conformance/subgroups/test_subgroup_ballot.cpp:
```diff
@@ -0,0 +1,1089 @@
+    "    XY(xy,gid);\n"
+    "    Type x = in[gid];\n"
+    "    uint4 value = (uint4)(10,0,0,0);\n"
+    "    if (get_sub_group_local_id() & 1) {"
```
The value returned by sub_group_inverse_ballot() does not depend on the uniformity of control flow within the subgroup. Why is divergence being tested here? It may be better not to test it without a clear idea as to the difference that it might contribute.

If it is necessary to test it, this should be done as per sub_group_elect() once the testing for that appropriately covers a maximum subgroup size of 128.

gwawiork replied:
It will by solved by https://github.com/KhronosGroup/OpenCL-CTS/issues/1205

StuartDBrady replied:
I think #1207 is the relevant issue for this comment.

StuartDBrady commented on test_conformance/subgroups/test_subgroup_ballot.cpp:
```diff
@@ -0,0 +1,1089 @@
+    "    Type x = in[gid];\n"
+    "    uint index = xy[gid].z;\n"
+    "    uint4 value = (uint4)(10,0,0,0);\n"
+    "    if (get_sub_group_local_id() & 1) {"
```
As per sub_group_ballot_inverse(), sub_group_ballot_bit_extract() is not sensitive to uniformity of control flow. Why is divergence tested here?

gwawiork replied:
Will be solved by https://github.com/KhronosGroup/OpenCL-CTS/issues/1207

StuartDBrady replied:
Agreed.

StuartDBrady commented:
> Thank you all for contribution and review. For today that is all I have to pushed out. Other issues will be address as separate GitHub Issues.

This still lacks changes relating to divergence discussed previously. These changes are needed for adequate coverage with a maximum subgroup size of 64 or more.

It may be the most pragmatic approach to merge this and raise follow-up issues for fixes, although we should do so with the understanding that we will be merging with large gaps in coverage.

StuartDBrady reviewed (approved):
This should be merged as-is so that the remaining issues can be addressed with follow-up PRs, given that reviewing the required changes in this PR is unworkable.

As stated, the testing is not yet in a satisfactory state, and so it is essential that this follow-up work be done.

gwawiork commented:
All comments are addressed by GitHub Issues in my opinion.

kpet commented:
Agreed in the 2021/04/06 teleconference: merging as is to enable forward progress. Further PRs will make required point fixes and keep improving the tests.
